### PR TITLE
BLD: Makefile, requirements[-test].txt, scripts/install_appenginesdk.py, docker/Dockerfile[.jinja], docker/Makefile, docker-compose.yml

### PR DESCRIPTION
**Pull request by westurner**
BLD: Makefile, requirements[-test].txt, scripts/install_appenginesdk.py, docker/Dockerfile[.jinja], docker/Makefile, docker-compose.yml

- schema.org *universal* build scripts
- [ ] Review (self)
- [ ] Review (____)

### CODE REVIEW

**Comment by westurner** [comment]
This is rebased out of https://github.com/schemaorg/schemaorg/pull/972 and mostly as yet untested (aside from hundreds of manual test invocations)

**Comment by westurner** [comment]
This is rebased out of https://github.com/schemaorg/schemaorg/pull/972 and mostly as yet untested (aside from hundreds of manual test invocations)

[edit]

- [ ] TST: docker/*
  - [ ] BLD: docker/Makefile: ``build_all`` (``make -C docker/ build_all``)


[edit]

Why test/wrap all of AppEngineSDK and the project build/test requirements for all of these OS when ``dev_appserver.py`` just works on Windows, OSX, Linux?

- Reproducibility:
  - Ideally, there is a Docker image containing the project/**app** source code (e.g. maybe initially with a *shallow clone* (``--depth=20``)) and any/all requisite build requirements (OS, OS packages, OS configuration, stack requirements (/**packages**? (see: FPM)), app requirements, app **package**, app configuration; such that the entire environment is reproducible at test time.

File: scripts/install_appenginesdk.py [hunk 1]
```diff
@@ -0,0 +1,1480 @@
...
+
+    @property
+    def success(self):
+        return all(step.success for step in self)
+
+    def __repr__(self):
+        return "BuildstepPlan([%s])" % json_dumps(self)
+
+
+# class BuildstepResultList(list):
```
Comment (westurner) [inline scripts/install_appenginesdk.py]: CLN: trim this out

File: scripts/install_appenginesdk.py [hunk 2]
```diff
@@ -0,0 +1,1480 @@
...
+            kwargs = step.data.get('kwargs', {})
+            func = step.data['func']
+            step.result = func(*args, **kwargs)
+            step.result.name = step.data['name']
+            yield step
+
+    def build_plan(self, cfg=None, steps=None):
+        return BuildstepPlan(self.build_plan_iter(cfg=cfg, steps=steps))
+
+    def build(self, cfg=None, steps=None):
```
Comment (westurner) [inline scripts/install_appenginesdk.py]: - [ ] CLN: trim this out

File: scripts/install_appenginesdk.py [hunk 3]
```diff
@@ -0,0 +1,1480 @@
...
+        if retcode != expectreturncode:
+            raise subprocess.CalledProcessError(
+                retcode,
+                [('cmd', cmd),
+                 ('args', args),
+                 ('kwargs', kwargs),
+                 ('retcode', retcode),
+                 ('expectreturncode', expectreturncode)])
+    return retcode
+    #return BuildstepResult(returncode=retcode)
```
Comment (westurner) [inline scripts/install_appenginesdk.py]: - [ ] CLN

File: scripts/install_appenginesdk.py [hunk 4]
```diff
@@ -0,0 +1,1480 @@
...
+        force (bool): if True, delete any existing file first
+        destfile (str): destination path to expect file to be downloaded to
+    Returns:
+        tuple: (success:bool, cmd:str)
+    """
+    if url is None:
+        raise ValueError((url, 'is None'))
+    if destdir is None:
+        raise ValueError((destdir, 'is None'))
+    # curl -S --continue - -O '$(APPENGINESDK_ARCHIVE_URL)'
```
Comment (westurner) [inline scripts/install_appenginesdk.py]: - [ ] DOC: to docstring

File: scripts/install_appenginesdk.py [hunk 5]
```diff
@@ -0,0 +1,1480 @@
...
+    'DEV_APPSERVER'
+]
+
+ENVIRON_ALL = APPENGINESDK_INSTALL_ENVIRON_KEYS
+ENVIRON_EXPORT_KEYS = APPENGINESDK_ENVIRON_KEYS
+
+
+def shell_escape_single(str_):
+    # TODO:
+    return str_.replace("'", """'"'"'""") # TODO TODO
```
Comment (westurner) [inline scripts/install_appenginesdk.py]: TODO: check/compare this w/ other implementations

File: scripts/install_appenginesdk.py [hunk 6]
```diff
@@ -0,0 +1,1480 @@
...
+                         ('APPENGINESDK_PREFIX',
+                          cfg['APPENGINESDK_PREFIX'])))
+    cmd = ('unzip', '-q',
+           cfg['APPENGINESDK_ARCHIVE_PATH'],
+           '-d', cfg['APPENGINESDK_BASEPATH'])
+    cwd = cfg['APPENGINESDK_ARCHIVE_DESTDIR']
+    retcode = subprocess_call_expect(
+        cmd,
+        cwd=cwd,
+        expectreturncode=0)  # XXX TODO
```
Comment (westurner) [inline scripts/install_appenginesdk.py]: - [ ] CLN

File: scripts/install_appenginesdk.py [hunk 7]
```diff
@@ -0,0 +1,1400 @@
...
+                    result = obj[1]
+                else:
+                    raise ValueError((type(obj), obj))
+                # TODO step._repr_checkbox_
+                if result is None:
+                    if obj:
+                        if hasattr(obj, '_repr_checkbox_'):
+                            yield indentstr + obj._repr_checkbox_()
+                        else:
+                            yield indentstr + '+ [ ] {!s}'.format(obj)
```
Comment (westurner) [inline scripts/install_appenginesdk.py]: - [ ] UBY: trace ``name`` through
  - a BuildstepResult may not have an explicit name defined (e.g. defaults to ``func.func_name``)
    - [x] explicit name='check'
    - [x] explicit name='install'

File: scripts/install_appenginesdk.py [hunk 8]
```diff
@@ -0,0 +1,1400 @@
...
+        log.debug(json_dumps(ctxt))
+        ctxt['result'] = result = f(*args, **kwargs)
+        if not isinstance(result, BuildstepResult):
+            raise ValueError((result, 'is not a BuildstepResult'))
+        if result.data.get('step') is None:
+            result.data['step'] = f
+        ctxt['name'] = f.func_name = result.name = (
+            kwargs.get("name",
+                       get_obj_name(f,
+                                    get_obj_name(result))))
```
Comment (westurner) [inline scripts/install_appenginesdk.py]: - [ ] does settings ``f.func_name`` here here any affect on test tracebacks?

File: scripts/install_appenginesdk.py [hunk 9]
```diff
@@ -0,0 +1,1400 @@
...
+            d.setdefault('name', get_obj_name(obj))
+            srcpath = inspect.getsourcefile(obj)
+            modname = obj.__module__
+            if modname == '__main__':
+                modname = ''
+            else:
+                modname = modname + '.'
+            x = "{}##{}{}".format(srcpath, modname, obj.__name__)
+
+            # NEW format specification minilanguages!!@#!
```
Comment (westurner) [inline scripts/install_appenginesdk.py]: - [ ] CLN

File: scripts/install_appenginesdk.py [hunk 10]
```diff
@@ -0,0 +1,1400 @@
...
+                        debug=None):
+        name = self.name if name is UNSET else name
+        success = self.success if success is UNSET else success
+        msg = self.data.get('msg') if msg is UNSET else msg
+        if debug is None:
+            debug = getattr(self, 'debug', None)
+        if name is None:
+            name = get_obj_name(self.data['step'], default=None)
+        if name is None:
+            raise ValueError(('name is  None', self))
```
Comment (westurner) [inline scripts/install_appenginesdk.py]: - [] TST,REF: cls.name 
  - BuildstepStep.result.name
  - BuildstepResult.name
  - BuildstepStep.name

File: scripts/install_appenginesdk.py [hunk 11]
```diff
@@ -0,0 +1,1400 @@
...
+    prs.add_option('--environ',
+                   dest='config_from_environ',
+                   action='store_true',
+                   help='Read environment variables from os.environ into cfg')
+
+    prs.add_option('--version',
+                   dest='APPENGINESDK_VERSION',
+                   action='store',
+                   default=None,
+                   help="APPENGINESDK_VERSION to install")
```
Comment (westurner) [inline scripts/install_appenginesdk.py]: - [ ] SEC,ENH: determine what the remote url for VERSION.yaml is from appenginesdk .py

**Comment by danbri** [comment]
It is hard to know what to make of this. There's clearly a lot of work here but it is hard to tell from the pull request what it all does exactly? I though initially it was to bring up a dev environment within the Docker framework, but perhaps there's also some larger notion of a schema.org project 'makefile' too?

**Comment by westurner** [comment]
* "Automate the build"
https://en.wikipedia.org/wiki/Continuous_integration#Automate_the_build

* TST/BLD scripts should exit with nonzero on error.

* AFAICT, there are not yet reference images for GAESDK.

* Because this project depends on GAESDK, it's not possible to test it
without installing GAESDK; so I wrote an installer script, in Python (such
that it could work in Windows, too), which accepts a version number as a
CLI parameter. (The Makefile also works alike)
  * It should not be necessary to unzip GAESDK for every automated build
(because this is a very simple WSGI application and there aren't many
actual dependencies on GAESDK).

* Neither the GAESDK docs nor the schemaorg docs list the actual package
dependencies. These Dockerfile list, ANAICT, the Linux package dependencies
for this project (and transitively GAESDK)

Why multiple Dockerfile for different distros?

* Pick one?
  * "I just don't do distro _____"

**Comment by danbri** [comment]
We're unlikely to merge this as-is (too much complexity for us, sorry) but @RichardWallis and I have discussed factoring out the appengine dependencies from sdoapp.py into a separate file, which should make it easier to run up test sites without GAE.

**Comment by westurner** [comment]
* Thanks!
* You may want to cherry-pick "TST,BLD: scripts/run_tests.py: return nonzero on error" https://github.com/westurner/schemaorg/commit/ee112620a4c6b689578b6ac26bf5c3c098beefc4

**Pull request closed without merging**